README.md
# CoCoKeys2USB
A hardware keyboard encoder for the TRS-80 Color Computer

v1.2 - initial commit

v1.21 - added switching between emulators with Shift+Enter to support CoCo1/2 keyboards

v1.22 - added Ctrl+'@' to emulate '@' keypress and Ctrl+'Left-Arrow' to emulate destructive backspace

v1.23 - added better support for OS-9 Ctrl+ keystrokes


CoCo_HID_Keyboard_v1.23_/change_log.txt
CoCo Keys2USB
*****************

v1.23 10-26-2021
-----------------
* added better support for OS-9 Ctrl Keys
* added cocoKeyIndex.h for key index lookups


v1.22 07-15-2020
-----------------
* added Ctrl+@ to emulate '@' on PC
* added Ctrl+Left-Arrow for destructive backspace on PC


v1.21 05-21-2020
-----------------
* added Shift+Enter to cycle key maps for CoCo1/2 keyboards


v1.2  04-20-2020
-----------------
* initial release

CoCo_HID_Keyboard_v1.22_/Keyboard.h
/*
  Keyboard.h

  Copyright (c) 2015, Arduino LLC
  Original code (pre-library): Copyright (c) 2011, Peter Barrett

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef KEYBOARD_h
#define KEYBOARD_h

#include "HID.h"

#if !defined(_USING_HID)

#warning "Using legacy HID core (non pluggable)"

#else

//================================================================================
//================================================================================
//  Keyboard

#define KEY_LEFT_CTRL     0x80
#define KEY_LEFT_SHIFT    0x81
#define KEY_LEFT_ALT      0x82
#define KEY_RIGHT_CTRL    0x84
#define KEY_RIGHT_SHIFT   0x85
#define KEY_RIGHT_ALT     0x86

#define KEY_UP_ARROW      0xDA
#define KEY_DOWN_ARROW    0xD9
#define KEY_LEFT_ARROW    0xD8
#define KEY_RIGHT_ARROW   0xD7
#define KEY_BACKSPACE     0xB2
#define KEY_TAB           0xB3
#define KEY_RETURN        0xB0
#define KEY_ESC           0xB1
#define KEY_INSERT        0xD1
#define KEY_DELETE        0xD4
#define KEY_PAGE_UP       0xD3
#define KEY_PAGE_DOWN     0xD6
#define KEY_HOME          0xD2
#define KEY_END           0xD5
#define KEY_CAPS_LOCK     0xC1
#define KEY_F1            0xC2
#define KEY_F2            0xC3
#define KEY_F3            0xC4
#define KEY_F4            0xC5
#define KEY_F5            0xC6
#define KEY_F6            0xC7
#define KEY_F7            0xC8
#define KEY_F8            0xC9
#define KEY_F9            0xCA
#define KEY_F10           0xCB
#define KEY_F11           0xCC
#define KEY_F12           0xCD
#define KEY_F13           0xF0
#define KEY_F14           0xF1
#define KEY_F15           0xF2
#define KEY_F16           0xF3
#define KEY_F17           0xF4
#define KEY_F18           0xF5
#define KEY_F19           0xF6
#define KEY_F20           0xF7
#define KEY_F21           0xF8
#define KEY_F22           0xF9
#define KEY_F23           0xFA
#define KEY_F24           0xFB


//  Low level key report: up to 6 keys and shift, ctrl etc at once
typedef struct
{
  uint8_t modifiers;
  uint8_t reserved;
  uint8_t keys[6];
} KeyReport;

class Keyboard_ : public Print
{
private:
  KeyReport _keyReport;
  void sendReport(KeyReport* keys);
public:
  Keyboard_(void);
  void begin(void);
  void end(void);
  size_t write(uint8_t k);
  size_t write(const uint8_t *buffer, size_t size);
  size_t press(uint8_t k);
  size_t release(uint8_t k);
  void releaseAll(void);
};
extern Keyboard_ Keyboard;

#endif
#endif


CoCo_HID_Keyboard_v1.23_/cocoKeyIndex.h
/*
    cocoKeyIndex.h v1.0
    by: Paul Fiscarelli
    
    Copyright (c) 2020, Paul Fiscarelli

    MIT License

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.
    
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/

//================================================================================
// Key indexes for CoCo key lookup values

#define KEY_AT_INDEX        0x00
#define KEY_A_INDEX         0x01
#define KEY_B_INDEX         0x02
#define KEY_C_INDEX         0x03
#define KEY_D_INDEX         0x04
#define KEY_E_INDEX         0x05
#define KEY_F_INDEX         0x06
#define KEY_G_INDEX         0x07
#define KEY_H_INDEX         0x08
#define KEY_I_INDEX         0x09
#define KEY_J_INDEX         0x0A
#define KEY_K_INDEX         0x0B
#define KEY_L_INDEX         0x0C
#define KEY_M_INDEX         0x0D
#define KEY_N_INDEX         0x0E
#define KEY_O_INDEX         0x0F
#define KEY_P_INDEX         0x10
#define KEY_Q_INDEX         0x11
#define KEY_R_INDEX         0x12
#define KEY_S_INDEX         0x13
#define KEY_T_INDEX         0x14
#define KEY_U_INDEX         0x15
#define KEY_V_INDEX         0x16
#define KEY_W_INDEX         0x17
#define KEY_X_INDEX         0x18
#define KEY_Y_INDEX         0x19
#define KEY_Z_INDEX         0x1A
#define KEY_UP_INDEX        0x1B
#define KEY_DOWN_INDEX      0x1C
#define KEY_LEFT_INDEX      0x1D
#define KEY_RIGHT_INDEX     0x1E
#define KEY_SPACE_INDEX     0x1F
#define KEY_0_INDEX         0x20
#define KEY_1_INDEX         0x21
#define KEY_2_INDEX         0x22
#define KEY_3_INDEX         0x23
#define KEY_4_INDEX         0x24
#define KEY_5_INDEX         0x25
#define KEY_6_INDEX         0x26
#define KEY_7_INDEX         0x27
#define KEY_8_INDEX         0x28
#define KEY_9_INDEX         0x29
#define KEY_COLON_INDEX     0x2A
#define KEY_SEMI_INDEX      0x2B
#define KEY_COMMA_INDEX     0x2C
#define KEY_HYPHEN_INDEX    0x2D
#define KEY_PERIOD_INDEX    0x2E
#define KEY_SLASH_INDEX     0x2F
#define KEY_ENTER_INDEX     0x30
#define KEY_CLEAR_INDEX     0x31
#define KEY_BREAK_INDEX     0x32
#define KEY_ALT_INDEX       0x33
#define KEY_CTRL_INDEX      0x34
#define KEY_F1_INDEX        0x35
#define KEY_F2_INDEX        0x36
#define KEY_SHIFT_INDEX     0x37


CoCo_HID_Keyboard_v1.23_/cocoKeyMaps.h
/*
    cocoKeyMaps.h v1.1
    by: Paul Fiscarelli
    
    Copyright (c) 2020, Paul Fiscarelli

    MIT License

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:
    
    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.
    
    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/
#include "Keyboard.h"

#define KEY_AT  0x5B
#define KEY_A   0x61
#define KEY_B   0x62
#define KEY_C   0x63
#define KEY_D   0x64
#define KEY_E   0x65
#define KEY_F   0x66
#define KEY_G   0x67
#define KEY_H   0x68
#define KEY_I   0x69
#define KEY_J   0x6A
#define KEY_K   0x6B
#define KEY_L   0x6C
#define KEY_M   0x6D
#define KEY_N   0x6E
#define KEY_O   0x6F
#define KEY_P   0x70
#define KEY_Q   0x71
#define KEY_R   0x72
#define KEY_S   0x73
#define KEY_T   0x74
#define KEY_U   0x75
#define KEY_V   0x76
#define KEY_W   0x77
#define KEY_X   0x78
#define KEY_Y   0x79
#define KEY_Z   0x7A
#define KEY_0   0x30
#define KEY_1   0x31
#define KEY_2   0x32
#define KEY_3   0x33
#define KEY_4   0x34
#define KEY_5   0x35
#define KEY_6   0x36
#define KEY_7   0x37
#define KEY_8   0x38
#define KEY_9   0x39

#define KEY_TAB         0x09
#define KEY_F1          0xC2
#define KEY_F2          0xC3
#define KEY_SHIFT       0x81
#define KEY_SPACE       0x20
#define KEY_COLON       0x2D
#define KEY_SEMICOLON   0x3B
#define KEYPAD_ASTERIX  0xDD
#define KEY_COMMA       0x2C
#define KEY_PERIOD      0x2E
#define KEY_SLASH       0x2F
#define KEY_MINUS       0x3D

uint8_t keyMapVCC[56] = {KEY_RIGHT_ALT, KEY_A, KEY_B, KEY_C, KEY_D, KEY_E, KEY_F, KEY_G,
                               KEY_H, KEY_I, KEY_J, KEY_K, KEY_L, KEY_M, KEY_N, KEY_O,
                               KEY_P, KEY_Q, KEY_R, KEY_S, KEY_T, KEY_U, KEY_V, KEY_W,
                               KEY_X, KEY_Y, KEY_Z, KEY_UP_ARROW, KEY_DOWN_ARROW, KEY_BACKSPACE, KEY_RIGHT_ARROW, KEY_SPACE,
                               KEY_0, KEY_1, KEY_2, KEY_3, KEY_4, KEY_5, KEY_6, KEY_7,
                               KEY_8, KEY_9, KEY_COLON, KEY_SEMICOLON, KEY_COMMA, KEY_MINUS, KEY_PERIOD, KEY_SLASH,
                               KEY_RETURN, KEY_HOME, KEY_END, KEY_LEFT_ALT, KEY_LEFT_CTRL, KEY_F1, KEY_F2, KEY_LEFT_SHIFT};

uint8_t keyMapXRoar[56] = {KEY_AT, KEY_A, KEY_B, KEY_C, KEY_D, KEY_E, KEY_F, KEY_G,
                                 KEY_H, KEY_I, KEY_J, KEY_K, KEY_L, KEY_M, KEY_N, KEY_O,
                                 KEY_P, KEY_Q, KEY_R, KEY_S, KEY_T, KEY_U, KEY_V, KEY_W,
                                 KEY_X, KEY_Y, KEY_Z, KEY_UP_ARROW, KEY_DOWN_ARROW, KEY_LEFT_ARROW, KEY_RIGHT_ARROW, KEY_SPACE,
                                 KEY_0, KEY_1, KEY_2, KEY_3, KEY_4, KEY_5, KEY_6, KEY_7,
                                 KEY_8, KEY_9, KEY_COLON, KEY_SEMICOLON, KEY_COMMA, KEY_MINUS, KEY_PERIOD, KEY_SLASH, 
                                 KEY_RETURN, KEY_HOME, KEY_ESC, KEY_LEFT_ALT, KEY_LEFT_CTRL, KEY_F1, KEY_F2, KEY_LEFT_SHIFT};

uint8_t keyMapMAME[56] = {KEY_AT, KEY_A, KEY_B, KEY_C, KEY_D, KEY_E, KEY_F, KEY_G,
                                KEY_H, KEY_I, KEY_J, KEY_K, KEY_L, KEY_M, KEY_N, KEY_O,
                                KEY_P, KEY_Q, KEY_R, KEY_S, KEY_T, KEY_U, KEY_V, KEY_W,
                                KEY_X, KEY_Y, KEY_Z, KEY_UP_ARROW, KEY_DOWN_ARROW, KEY_LEFT_ARROW, KEY_RIGHT_ARROW, KEY_SPACE,
                                KEY_0, KEY_1, KEY_2, KEY_3, KEY_4, KEY_5, KEY_6, KEY_7,
                                KEY_8, KEY_9, KEY_COLON, KEY_SEMICOLON, KEY_COMMA, KEY_MINUS, KEY_PERIOD, KEY_SLASH,
                                KEY_RETURN, KEY_HOME, KEY_ESC, KEY_LEFT_ALT, KEY_LEFT_CTRL, KEY_F1, KEY_F2, KEY_LEFT_SHIFT};                                 
